### PR TITLE
Implement Madry's Protocol as adversarial trainer

### LINKED ISSUE #294: Implement Madry's Protocol as adversarial training example
Create API for adversarial trainers which train classifiers on adversarial examples.

Implement Madry's Protocol as a new example (paper link: https://arxiv.org/abs/1706.06083)

### PR DESCRIPTION
**Pull request by beat-buesser**
Create API for adversarial trainers which train classifiers on adversarial examples and implement Madry's Protocol as a new example (https://arxiv.org/abs/1706.06083)

Fixes #294

### CODE REVIEW

File: art/defences/trainer/adversarial_trainier_madry_pgd.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements adversarial training following Madry's Protocol.
```
Comment (mathsinn) [inline art/defences/trainer/adversarial_trainier_madry_pgd.py]: There is a typo in the module name ("trainier" -> "trainer")

File: art/defences/trainer/adversarial_trainier_madry_pgd.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        :param max_iter: The maximum number of iterations.
+        :type max_iter: `int`
+        :param num_random_init: Number of random initialisations within the epsilon ball. For num_random_init=0
+            starting at the original input.
+        :type num_random_init: `int`
+        """
+
+        self.max_num_training_steps = max_num_training_steps
+        self.batch_size = batch_size
+        self.nb_epochs = int(max_num_training_steps / (50000 / batch_size))
```
Comment (mathsinn) [inline art/defences/trainer/adversarial_trainier_madry_pgd.py]: Where does the `50000` come from? Supposedly this is the length of the training set. Perhaps it's easier for the user to directly provide the number of epochs, or is `max_num_training_steps` used anywhere else?

File: art/defences/trainer/adversarial_trainier_madry_pgd.py (hunk 1 above)
Comment (beat-buesser) [reply art/defences/trainer/adversarial_trainier_madry_pgd.py]: Fixed with 89a0041fa507d3300be7aba7d28471fe22c16e1d

File: art/defences/trainer/adversarial_trainier_madry_pgd.py (hunk 2 above)
Comment (beat-buesser) [reply art/defences/trainer/adversarial_trainier_madry_pgd.py]: I think the 50000 came with the source code. After our discussion I replaced it with the argument `nb_epochs` in 661c27dfd27cb4c079dbf0959e8b1700e35891b9.

**Comment by beat-buesser** [comment]
@mathsinn I have fixed all requests and added a unittest.

**Pull request merged**